Document required client changes for Rekor v2 (#255)
sigstore/rekor-tiles: CLIENTS.md

            "keyDetails": "PKIX_ECDSA_P256_SHA_256" # Must match signing algorithm
        }
    }
}
```

The response will be a `TransparencyLogEntry` message, which should be persisted
in a bundle. Clients no longer need to transform the Rekor response into a `TLE`
message to store in the bundle.

## Handling Longer Requests

Clients need to increase request timeouts to at least 10 seconds.

Rekor now batches uploads so that checkpoints are published less frequently.
Additionally, we plan to support synchronous witnessing, where third-party
witnesses independently verify the consistency of the log and Rekor provides
co-signed checkpoints with each upload response.

## Signed RFC 3161 Timestamps

Rekor will no longer return SignedEntryTimestamps or include integrated time
in the response. Clients must fetch an RFC 3161 signed timestamp from a trusted
timestamp authority and include the signed timestamp in the bundle.

Sigstore now operates a timestamp authority at `timestamp.sigstore.dev` and
`timestamp.sigstage.dev` for staging, and the roots for these services will
be included in the TrustedRoot distributed via TUF. Clients may request timestamps

> from other trusted timestamp authorities as well.

Reviewer: I assume we are expecting clients that use other "trusted timestamp" authorities to inject that information into the trusted root? Client code should not fetch the timestamp cert from the provider

> ## SigningConfig support

Author: @jku I added some more info about the SigningConfig and motivation behind the changes, lemme know if you have more questions that I can clarify in this doc

# Client Changes for Rekor v2

This document outlines the changes clients need to make to support
Rekor v2.

## Rekor v2 API

Rekor v2 supports HTTP or gRPC, like Fulcio. For Go, we have implemented a client

> already. For other languages, they can either use the OpenAPI docs or generated
> gRPC service proto, or create their own client.

Reviewer: ```suggestion
already. For other languages, they can either use the [OpenAPI docs](https://github.com/sigstore/rekor-tiles/tree/main/docs/openapi), [gRPC service proto](https://github.com/sigstore/rekor-tiles/tree/main/api/proto), or create their own client.
```

The service implements one write API, `/api/v2/log/entries`. Example JSON request bodies
below:

> ```json

Reviewer: I haven't tested this, but `jsonc` appears to be a markdown rendering formation for json with comments.

// Request with a Fulcio certificate
{
    "hashedRekordRequestV0_0_2": {
        "digest": "<base64 digest of artifact>", # Must use hash algorithm from key_details
        "signature": {
            "content": "<base64 signature>",
            "verifier": {
                "x509Certificate": "<base64 DER-encoded certificate>",
                "keyDetails": "PKIX_ECDSA_P256_SHA_256" # Must match signing algorithm
            }
        }
    }
}

// Request with a self-managed key
{
    "hashedRekordRequestV0_0_2": {
        "digest": "<base64 digest of artifact>",
        "signature": {
            "content": "<base64 signature>",
            "verifier": {
                "publicKey": {

>                     "raw_bytes": "<base64 DER-encoded public key>"

Reviewer 2: ```suggestion
                    "rawBytes": "<base64 DER-encoded public key>"
```

                },
                "keyDetails": "PKIX_ECDSA_P256_SHA_256"
            }
        }
    }
}

// Request with an attestation
{
    "dsseRequestV0_0_2": {
        "envelope": {
            "payload": "<base64-encoded message>",
            "payloadType": "<type, e.g. application/vnd.in-toto+json>",
            "signatures": [
                {
                    "sig": "<base64-encoded signature>",
                    "keyid": ""
                }
            ]
        },
        "verifier": {

>             "x509Certificate": "<base64 DER-encoded certificate>", # May also be a public key

Reviewer 2: Would be clearer if the example was repeated like you are for hashedrekord, with the `publicKey` field instead of `x509Certificate`

            "keyDetails": "PKIX_ECDSA_P256_SHA_256" # Must match signing algorithm
        }
    }
}
```

The response will be a `TransparencyLogEntry` message, which should be persisted
in a bundle. Clients no longer need to transform the Rekor response into a `TLE`
message to store in the bundle.

## Handling Longer Requests

Clients need to increase request timeouts to at least 10 seconds.

Rekor now batches uploads so that checkpoints are published less frequently.
Additionally, we plan to support synchronous witnessing, where third-party
witnesses independently verify the consistency of the log and Rekor provides

> co-signed checkpoints with each upload response.

Author: I'll add the line about uploading in parallel, but the client shouldn't assume the same checkpoint for each, as Tessera might trigger a new checkpoint creation in the middle of the set of uploads. If a client needs to save space, it would need to request the latest checkpoint and compute inclusion proofs for all entries, but once we introduce synchronous witnessing, this might be an issue unless witnesses also add support for requesting previously co-signed checkpoints.

Clients must be given the inclusion proof and checkpoint for an entry,
which must be stored in a bundle.

This should have no impact on clients as inclusion proofs were already
required in bundles. This will only impact monitors, and the read API
changes are detailed below.

## C2SP Checkpoints

The checkpoints the log provides will conform to the
[C2SP checkpoint spec](https://github.com/C2SP/C2SP/blob/main/tlog-checkpoint.md).
Clients must check that their checkpoint verification implementation properly
handles these checkpoints, which could include:

* Multiple signatures
* Optional extension lines
* Updated key names, which will match the shard URL
* Key IDs calculated per the [signed note spec](https://github.com/C2SP/C2SP/blob/main/signed-note.md#signatures)
  * ECDSA and Ed25519 key IDs are based on the spec
  * For RSA, the signature type is `0xFF` and we append `PKIX-RSA-PKCS#1v1.5`,
    `key ID = SHA-256(key name || 0x0A || 0xFF || PKIX-RSA-PKCS#1v1.5 || public key)[:4]`

## TrustedRoot With Multiple Logs

Clients must verify that verification works with a TrustedRoot
with multiple logs with overlapping validity windows. Confirm that

> the client will fetch the correct verification key from the TrustedRoot.

Reviewer: Fetch correct key by matching logid/validity window?

Author: Added. I'd say we use log ID for now. We should later align log ID with how we're computing log IDs for checkpoints.

## Monitoring/Auditing

One of the more significant changes in a tile-backed log is the changes
to the read API. Inclusion and consistency proofs are not served via an
API, rather the client requests the set of tiles necessary to compute
the inclusion or consistency proof.

When monitoring the log searching for entries, the monitor will not request
entries by index, but by tile. Monitors can choose to only fetch complete

> tiles or request partial tiles. It is recommended to request partial tiles,

Reviewer 2: It might not be clear to someone brand-new to this what a partial tile is or why you would want it. Maybe add some clarification that the latest tile is generally a partial tile.

                    "keyid": ""
                }
            ]
        },
        "verifier": {
            "x509Certificate": "<base64 DER-encoded certificate>",
            // Must match signing algorithm
            "keyDetails": "PKIX_ECDSA_P256_SHA_256"
        }
    }
}

// Request with an attestation with a self-managed key
{
    "dsseRequestV0_0_2": {
        "envelope": {
            "payload": "<base64-encoded message>",
            "payloadType": "<type, e.g. application/vnd.in-toto+json>",
            "signatures": [
                {
                    "sig": "<base64-encoded signature>",
                    "keyid": ""
                }
            ]
        },
        "verifier": {
            "publicKey": {
                "rawBytes": "<base64 DER-encoded public key>"
            },
            // Must match signing algorithm
            "keyDetails": "PKIX_ECDSA_P256_SHA_256"
        }
    }
}
```

The response will be a
[`TransparencyLogEntry` message](https://github.com/sigstore/protobuf-specs/blob/5296f13d62e7fad428581d969f664c30cc52f549/protos/sigstore_rekor.proto#L94),
which should be persisted in a bundle. Clients no longer need to transform the

> Rekor response into a `TLE` message to store in the bundle.

Reviewer 3: It would be helpful to show the shapes of the old and new responses.

### Two Entry Types

Rekor v2 only supports `hashedrekord` (`HashedRekordRequestV0_0_2`) and
`dsse` (`DSSERequestV0_0_2`) entry types, dropping a number of unused types
such as `jar`, `alpine`, `rpm`, and the older types `rekord` and `intoto`.
Additional types may be added in the future if there is demand, but this
will require updating the client specification so that all clients implement
support for these types.

### Certificate and Public Key Verifiers

Rekor v2 only supports signature verification using a certificate or a
public key, dropping support for PGP, minisign, pkcs7, SSH and TUF.
Additional verifiers may be added in the future, but this will also require
updating the client specification.

### Handling Longer Requests

Clients need to increase request timeouts to at least 10 seconds.

> Rekor now batches uploads so that checkpoints are published less frequently.

Reviewer 2: Not that this matters to clients but this is technically not true, the checkpoint publishing interval is not related to the batch interval.

Author: Thanks, removed the mention of batching and clarified that this is just about blocking until a checkpoint has been published.

Additionally, we plan to support synchronous witnessing, where third-party
witnesses independently verify the consistency of the log and Rekor provides
co-signed checkpoints with each upload response.

If a client needs to create multiple entries, it is recommended to upload those
entries in parallel.

## TrustedRoot lookup by checkpoint key ID rather than log ID

Log ID, the SHA-256 digest of the log's public key, is used as a "unique"
identifier for a log. The TransparencyLogEntry message includes it in a
bundle, and clients use that log ID to lookup the correct log public key
in the TrustedRoot to verify the bundle.

Rekor will no longer include log IDs in the response, and instead clients
should use the checkpoint key ID as specified in the
[C2SP spec](https://github.com/C2SP/C2SP/blob/main/signed-note.md#signatures)

> to lookup the correct log public key to verify a bundle. Each transparency

Reviewer 4: I think this might not be clear enough like this. How exactly do clients calculate the checkpoint key id for an entry -- is entrys checkpoint origin the "key name"?
 
 is there an example implementation of this?

Author: Added selection algorithms for this, thanks. The `TransparencyLogInstance` proto has been updated as well to document this.

tl;dr either 1) compare the bundle's `log_id` to the trust root log instance's `checkpoint_key_id` or `log_id` (what most clients are doing already), 2) enumerate all log instances and compare the checkpoint's key ID from the signature line to the first 4 bytes of the trust root log instance's `checkpoint_key_id`/`log_id` (which is what sigstore-python is doing), or 3) compute the log instance's checkpoint key ID using the instance's public key and origin and compare (this is Hard Mode, and I doubt any client will)

log instance in the TrustedRoot will include a `checkpoint_key_id` instead
of a `log_id`.

For more information, log IDs are not necessarily unique identifiers for

> a log, since a log may reuse its public key among instances. Additionally,

Reviewer 2: Why would we not just create a new key for the new instance?

Author: Added a line that for the public deployment, we always will. Private deployments may reuse keys.

the log origin is not necessarily a unique identifier, because multiple logs

> may be hosted by one origin. Even the combination of both is not necessarily

Reviewer 2: Each shard should be uniquely named

Author: Thanks, added a mention that this SHOULD NOT happen. Technically, a deployment can reuse the key and/or origin, which the checkpoint key ID calculation takes into account.

    // Fulcio service URLs.
    // Clients must select the first service from the list whose validity window
    // is active and the API version is supported by the client. Clients must
    // select the highest supported API version.
    // Clients can assume that this list is sorted from most recent to oldest.
    "caUrls": [
        {
            "url": "https://fulcio.sigstore.dev",
            "majorApiVersion": 1,
            "validFor": {
                // When clients should start using this service
                "start": "<UTC timestamp>",
                // Optional, when a service is turned down and clients should
                // treat the service as offline
                "end": "<UTC timestamp>"
            },
        }
    ],

    // OIDC service URLs
    "oidcUrls": [
        {
            "url": "https://oauth2.sigstore.dev/auth",
            "majorApiVersion": 1,
            "validFor": {
                "start": "<UTC timestamp>"
            },
        }
    ],

    // Rekor service URLs. This example shows multiple active logs.
    // The client should select the log with the highest API version
    // it supports.
    "rekorTlogUrls": [
        {

>             "url": "https://rekor-2025-1.sigstore.dev",

Reviewer 2: ```suggestion
            "url": "https://log2025-01.rekor.sigstore.dev",
```

>             "url": "https://rekor-2025-1.sigstore.dev",

Reviewer 2: ```suggestion
            "url": "https://log2025-01.rekor.sigstore.dev",
```

There are three possible scenarios for how clients should use kind and version
to parse an entry:

1. With a verification bundle with `canonicalized_body`, clients can use the
   bundle's `kind` and `version` or `canonicalized_body.kind` and
   `canonicalized_body.api_version` to parse `canonicalized_body.spec`.
2. Clients that do not persist `canonicalized_body` must reconstruct the body
   in order to compute the leaf hash to verify the inclusion proof. Clients can
   use the bundle's `kind` and `version` to reconstruct the log entry given
   the artifact. Remember to JSON-canonicalize the entry before computing the leaf hash.
3. Monitors reading the entry from a tiled entry bundle will not have the verification
   bundle, but will have the canonicalized body. A monitor must use the `kind` and
   `apiVersion` to parse `spec`.

### Certificate and Public Key Verifiers

Rekor v2 only supports signature verification using a certificate or a
public key, dropping support for PGP, minisign, pkcs7, SSH and TUF.
Additional verifiers may be added in the future, but this will also require
updating the client specification.

### Handling Longer Requests

> Clients need to increase request timeouts to at least 10 seconds.

Reviewer 3: ```suggestion
Clients need to increase request timeouts to at least 10 seconds when creating entries.
```

Author: Added, and I've doubled the timeout recommendation to 20 seconds.

```

Post an entry:

```bash
curl  -H \
 "Accept: application/json" -X \
POST http://localhost:3003/api/v2/log/entries -o rekor_response -d "{ \"hashedRekordRequestV0_0_2\":{ \"digest\":\"$(cat artifact_dgst|base64)\", \"signature\":{ \"content\": \"$(cat artifact.sig | base64)\", \"verifier\": { \"key_details\": \"PKIX_ECDSA_P256_SHA_256\", \"public_key\": { \"raw_bytes\": \"$(openssl base64 -d -in ec_public.pem | base64)\" } } } } }"
```

View the response with `cat rekor_response | jq .` Example:

```json
{
  "logIndex": "0",
  "logId": {
    "keyId": "2AtEIMfG6Y41yK0tcwRTBS2tjhOrjKGIpDkHFgp65g0="
  },
  "kindVersion": {
    "kind": "hashedrekord",
    "version": "0.0.2"
  },
  "integratedTime": "0",
  "inclusionPromise": null,
  "inclusionProof": {
    "logIndex": "0",
    "rootHash": "OWI5MjU1MWIxZTA0NGIzNmQ0ZGE0MTU0M2UwNjEwZjVhNzQyMDNiN2JiOWEzOWYwNDExOTZlMTI2MTg5NmM3NQ==",
    "treeSize": "1",
    "hashes": [],
    "checkpoint": {
      "envelope": "rekor-local\n1\nm5JVGx4ESzbU2kFUPgYQ9adCA7e7mjnwQRluEmGJbHU=\n\n— rekor-local 2AtEIIwnbtxrneJ7L1lQebfBRl7TxK84DTmx+kcZi7A25cBDgESI23f9ylThAlOireJ7U+H8eZF/4kJQcn9o5Qt8mQU=\n"
    }
  },

>   "canonicalizedBody": "eyJhcGlWZXJzaW9uIjoiMC4wLjIiLCJraW5kIjoiaGFzaGVkcmVrb3JkIiwic3BlYyI6eyJoYXNoZWRSZWtvcmRWMF8wXzIiOnsiZGF0YSI6eyJhbGdvcml0aG0iOiJTSEEyXzI1NiIsImRpZ2VzdCI6ImR5ajRlZG5ZSGpONC96c2pqQmVlTGFoUzlzbHA5N1o2N0xUQVZ4anJqWHc9In0sInNpZ25hdHVyZSI6eyJjb250ZW50IjoiTUVRQ0lCK1lQYTlvM1NOMHNRNHVkdUdmK21aeHdGZk9oRlowQ2d5K3A3VnQxbzJTQWlBUEZESHFPQUpMWW12dENXT3NEeU5ZMUg0VjN6bTRORURZczNOeXZIaDFQZz09IiwidmVyaWZpZXIiOnsia2V5RGV0YWlscyI6IlBLSVhfRUNEU0FfUDI1Nl9TSEFfMjU2IiwicHVibGljS2V5Ijp7InJhd0J5dGVzIjoiTUZrd0V3WUhLb1pJemowQ0FRWUlLb1pJemowREFRY0RRZ0FFMnNsT2Y4ZVpjajJtb1cydDRVRmo3dkNMNlFwRHprRHFxU1VtbTRPSkNWdklhdUtMeG0wYUdzM1ZNUFBmYXVNUGFNdXRuMC9zM2pnMHJyb0Z4b2ljeWc9PSJ9fX19fX0="

Reviewer 3: Also show the decoded canonicalized body.

Author: There is now a section on parsing the canonicalized body.
